Content.Server/Implants/SubdermalImplantSystem.cs

```diff
@@ -107,42 +111,84 @@ private void OnScramImplant(EntityUid uid, SubdermalImplantComponent component,
             _pullingSystem.TryStopPull(ent, pull);
 
         var xform = Transform(ent);
-        var entityCoords = xform.Coordinates.ToMap(EntityManager, _xform);
+        var targetCoords = SelectRandomTileInRange(xform, implant.TeleportRadius);
 
-        // try to find a valid position to teleport to, teleport to whatever works if we can't
-        var targetCoords = new MapCoordinates();
-        for (var i = 0; i < implant.TeleportAttempts; i++)
+        if (targetCoords != null)
         {
-            var distance = implant.TeleportRadius * MathF.Sqrt(_random.NextFloat()); // to get an uniform distribution
-            targetCoords = entityCoords.Offset(_random.NextAngle().ToVec() * distance);
+            _xform.SetWorldPosition(ent, targetCoords.Value.Position);
+            _xform.AttachToGridOrMap(ent, xform);
+            _audio.PlayPvs(implant.TeleportSound, ent);
+            args.Handled = true;
+        }
+    }
 
-            // prefer teleporting to grids
-            if (!_mapManager.TryFindGridAt(targetCoords, out var gridUid, out var grid))
-                continue;
+    private MapCoordinates? SelectRandomTileInRange(TransformComponent userXform, float radius)
+    {
+        var userCoords = userXform.Coordinates.ToMap(EntityManager, _xform);
+        var grids = _lookupSystem.GetEntitiesInRange<MapGridComponent>(userCoords, radius).ToList();
+        _random.Shuffle(grids);
 
-            // the implant user probably does not want to be in your walls
-            var valid = true;
-            foreach (var entity in grid.GetAnchoredEntities(targetCoords))
+        // Give preference to the grid the entity is currently on.
```

Comment:
Isn't there an overload of getentitiesinrange that returns a hashset? Alternatively you should just be storing this yourself and passing it in.

Reply:
I have no idea what you mean by this. It returns a hasset and I need to turn this into an array for me to actually shuffle the values. Do you want me to just PickAndTake over the hashset?

Reply:
Either that or storing a field for Hashset<Entity<MapGridComponent>> on the systems and clearing it before you need it.